### PR TITLE
SAMZA-2789: Remove cap on intermediate stream partition count for stream mode

### PR DESCRIPTION
**Pull request by mynameborat**
**Problem**: Intermediate stream partition count inference logic caps the partition size to 256 resulting in imbalances in work assignments to tasks

**Description**: As part of the intermediate partition size inference logic, we currently employ the following algorithm.

- partitionCount = Math.max(maxPartitionSize(inputStreams), maxPartitionSize(outputStreams))
- cap the partitionCount to MAX_INFERRED_PARTITIONS defined in the `IntermediateStreamManager` which is 256
- apply the inferred partition count to intermediate streams whose partition count is uninitialized

The logic above always caps the partition size of intermediate streams to 256 for all auto-created intermediate streams. This can prevent the job from scaling up uniformly as the intermediate partition assignment is capped to 256 tasks thereby rendering other tasks imbalanced in case of number tasks > 256.

**Changes**:
- Apply the cap only for batch mode as 256 limit was introduced for batch mode where number of files (partition) could be large
- Add unit tests for `IntermediateStreamManager`
- Minor java doc fix for `DefaultTaskExecutorFactory`

**Tests**: Added unit tests for the code changes

**API Changes**: None

**Upgrade Instructions**:
- Jobs that are temporarily worked around this constraint by setting `job.intermediate.stream.partitions` should remove the configuration in order for samza to infer and apply the partition count as described above
- Jobs that don't use `job.intermediate.stream.partitions` need no changes.

**Usage Instructions**: Refer to upgrade instruction.

### CODE REVIEW

**Comment by alnzng** [comment]
LGTM, thanks for making this change.

File: samza-core/src/main/java/org/apache/samza/task/DefaultTaskExecutorFactory.java [hunk 1]
```diff
@@ -50,9 +50,8 @@ public ExecutorService getTaskExecutor(Config config) {
    * {@inheritDoc}
    *
    * The choice of thread pool is determined based on the following logic
-   *    If job.operator.thread.pool.enabled,
-   *     a. Use {@link #getTaskExecutor(Config)} if job.container.thread.pool.size &gt; 1
-   *     b. Use default single threaded pool otherwise
+   *     1. Use {@link #getTaskExecutor(Config)} if job.container.thread.pool.size &gt; 1
```
Comment (ajothomas) [inline samza-core/src/main/java/org/apache/samza/task/DefaultTaskExecutorFactory.java:53]: Nit: This change seems unrelated and doesn't affect build. Do we need this with this PR?

**Review by ajothomas — CHANGES_REQUESTED** [review]
Looks good. just a minor comment

File: samza-core/src/test/java/org/apache/samza/execution/TestIntermediateStreamManager.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import com.google.common.collect.ImmutableSet;
+import org.apache.samza.config.ApplicationConfig.ApplicationMode;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.JobConfig;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+
+import org.mockito.MockitoAnnotations;
+import static org.mockito.Mockito.*;
```
Comment (dxichen) [inline samza-core/src/test/java/org/apache/samza/execution/TestIntermediateStreamManager.java:30]: Remove wildcard imports

**Review by dxichen — CHANGES_REQUESTED** [review]
Minor nit - thanks for the quick patch

File: samza-core/src/main/java/org/apache/samza/task/DefaultTaskExecutorFactory.java (hunk 1 above)
Comment (mynameborat) [reply samza-core/src/main/java/org/apache/samza/task/DefaultTaskExecutorFactory.java:53]: Yes. but I didn't want to open a separate PR for this one line java doc change.

File: samza-core/src/test/java/org/apache/samza/execution/TestIntermediateStreamManager.java (hunk 2 above)
Comment (mynameborat) [reply samza-core/src/test/java/org/apache/samza/execution/TestIntermediateStreamManager.java:30]: The convention in the code base is to allow wildcards only for static imports. Agree on non-wildcard for non-static imports.

**Review by ajothomas — APPROVED** [review]
